src/jdk.jlink/share/classes/jdk/tools/jlink/internal/JlinkTask.java

```diff
@@ -391,10 +410,99 @@ private JlinkConfiguration initJlinkConfig() throws BadArgs {
+    }
+
+    private static record LinkableRuntimesResult(boolean isLinkFromRuntime,
+                                                 ModuleFinder finder) {}
+
+    /**
+     * Perform sanity checks and setup actions for JDK linkable runtime links
+     *
+     * @param finder The module finder from packaged modules
+     * @param roots The roots (if any) of the modules.
+     * @return A module finder possibly adding the system modules for JDK
+     *         linkable runtimes if and only if java.base wasn't in the provided
+     *         module finder already. Otherwise returns the provided finder.
+     *         In both cases finders get the set of limited modules applied if
+     *         so requested with --limit-modules.
+     *
+     * @throws BadArgs if the jlink input options contain an invalid
+     *                 combination.
+     */
+    private LinkableRuntimesResult linkableJDKRuntimesInit(ModuleFinder finder,
+                                                           Set<String> roots)
+        throws BadArgs {
+        // Linkable JDK runtimes and an empty module path in conjunction
+        // with --keep-packaged-modules doesn't make sense as we are not linking
+        // from packaged modules to begin with.
+        if (options.modulePath.isEmpty() && options.packagedModulesPath != null) {
+            throw taskHelper.newBadArgs("err.runtime.link.packaged.mods");
+        }
+        boolean isLinkFromRuntime = options.modulePath.isEmpty();
+        // In case of custom modules outside the JDK we may
+        // have a non-empty module path, which must not include
+        // java.base. If it did, we link using packaged modules from that
+        // module path. If the module path does not include java.base, we must
+        // have a linkable JDK runtime. In that case we take the JDK modules
+        // from the run-time image.
+        if (finder.find("java.base").isEmpty()) {
+            isLinkFromRuntime = true;
+            ModuleFinder runtimeImageFinder = ModuleFinder.ofSystem();
+            finder = combinedFinders(finder, runtimeImageFinder, options.limitMods, roots);
+        }
+        return new LinkableRuntimesResult(isLinkFromRuntime, finder);
+    }
+
+    /**
+     * Creates a combined module finder of {@code finder} and
+     * {@code runtimeImageFinder} that first looks-up modules in the
+     * {@code runtimeImageFinder} and if not present in {@code finder}.
+     *
+     * @param finder A module finder based on packaged modules.
+     * @param runtimeImageFinder A system modules finder.
+     * @param limitMods The set of limited modules for the resulting
+     *                  finder (if any).
+     * @param roots All module roots.
+     *
+     * @return A combined finder, or the input finder, potentially applying
+     *         module limits.
+     */
+    private ModuleFinder combinedFinders(ModuleFinder finder,
+            ModuleFinder runtimeImageFinder, Set<String> limitMods,
```

Comment:
suggest to make `runtimeImageFinder` as the first parameter and `finder` second to match the order of the lookup.